databox/client/interfaces/IFunctionPlugin.ts

```diff
@@ -0,0 +1,18 @@
+import IFunctionExecOptions from './IFunctionExecOptions';
+import IFunctionSchema from './IFunctionSchema';
+import IFunctionContext from './IFunctionContext';
+import FunctionInternal from '../lib/FunctionInternal';
+
+export default interface IFunctionPlugin<
+  ISchema extends IFunctionSchema,
+  IOptions extends IFunctionExecOptions<ISchema> = IFunctionExecOptions<ISchema>,
+  IContext extends IFunctionContext<ISchema> = IFunctionContext<ISchema>,
+> {
+  name: string;
+  version: string;
+  shouldRun?: boolean;
+  onStart?(functionInternal: FunctionInternal<ISchema, IOptions>): void | Promise<void>;
+  beforeRun?(context: IContext): void | Promise<void>;
+  beforeClose?(): void | Promise<void>;
+  onClose?(): void | Promise<void>;
```

Comment:
beforeClose and onClose sound like the same thing.

Reply:
It was called "onBeforeClose" before. I agree this plugin event naming feels like they are all doing the same events. onExec, onRun, onBeforeClose, onClose. I think it all simplifies with the new plugin pattern we talked about, right?

Reply:
Yeah, the new pattern might fix it.

Dropping the "on" from "onBeforeClose" seems to make it even more confusing, but I guess this is just a difference of opinion.